### PR TITLE
Example: Sequence to sequence learning for addition using RNNs

### PR DESCRIPTION
**Pull request by Smerity**
An implementation of sequence to sequence learning for performing addition.
Input: "535+61"
Output: "596"
Padding is handled by using a repeated sentinel character (space).

By default, the JZS1 recurrent neural network is used. JZS1 was an "evolved" recurrent neural network that explicitly performed well on arithmetic benchmark in [An Empirical Exploration of Recurrent Network Architectures](http://jmlr.org/proceedings/papers/v37/jozefowicz15.pdf)

Input may optionally be inverted, shown to increase performance in many tasks in [Learning to Execute](http://arxiv.org/abs/1410.4615) and [Sequence to Sequence Learning with Neural Networks](http://papers.nips.cc/paper/5346-sequence-to-sequence-learning-with-neural-networks.pdf) theoretically as it introduces shorter term dependencies between source and target.

---

The task loosely follows the one set forth in [Learning to Execute](http://arxiv.org/abs/1410.4615) where training is continued until 99% accuracy is achieved. They use the teacher method (correct previous digit is provided as input even if the RNN gave the wrong digit) whilst our example doesn't use such help. The [Learning to Execute](http://arxiv.org/abs/1410.4615) paper also goes from length 5 to 65 but I didn't have enough computing power to perform such experiments.

Two digits inverted:
+ One layer JZS1 (128 HN) with 55 iterations = 99% train/test accuracy

Three digits inverted:
+ One layer JZS1 (128 HN) with 19 iterations = 99% train/test accuracy

Four digits inverted:
+ One layer JZS1 (128 HN) with 20 iterations = 99% train/test accuracy

Five digits inverted:
+ One layer JZS1 (128 HN) with 28 iterations = 99% train/test accuracy

---

After each training iteration, examples of the output are shown to help understand errors. Generally the errors are off by ones due to carrying.

    Q 188+273
    T 461 
    ☑ 461 
    
    Q 542+57 
    T 599 
    ☒ 699

### CODE REVIEW

**Comment by fchollet** [comment]
I'm somewhat surprised at how many training examples are generated. The training data represents a very significant fraction of all possible additions (in the 3 digits case: 30% of the entire space). But surely learning to add two 3-digits could be achieved with much fewer examples? At least elementary school kids don't need to memorize hundreds of thousands of examples to learn addition ;-)

I think comparable accuracy could be achieved with 3-5 times fewer examples, which would be more impressive (while less impressive than a kid).

One oddity I'm noticing: although validation accuracy climbs quite high (in my current test, 60% right now), close to 100% (at best 90%) of random examples shown are failures. I wonder what's going on?

**Comment by Smerity** [comment]
For the 3 digits case, you're right - 30% of the entire space is a really good chunk, especially given most of the remaining space is filtered out! (i.e. removing symmetry by only allowing A+B and not B+A)

Decreasing the default training size would do fine too, especially if it's a better first-run experience. The selected number works well for DIGITS = 2-5 and doesn't run too slowly on my CPU. Sadly the papers don't state how many samples were used to generate their results but once you're past 10 or so digits you have no lack of training space.

Re: oddity / validation accuracy, I believe it's performing accuracy over each of the characters in the output sequence. As such, given 100 examples where the output is a 4 digit number (3 digit + 3 digit), if it's hitting 60% accuracy we'd expect ~160 of the characters are incorrect. If those incorrect characters are evenly distributed, essentially all the examples would be incorrect - what you're seeing.

Accuracy is evaluated in a per character sense by default in this setup which matches the evaluation method in [Learning to Execute](http://arxiv.org/pdf/1410.4615v3.pdf). The paper also features "teacher forcing" where they tell the RNN what the previous digit was even if the RNN got it incorrect however. This code doesn't follow that approach as (a) it would muddy the encoder / decoder example and (b) this method is easier to modify to real world tasks where we don't know the true answer.

**Comment by fchollet** [comment]
Cool, makes sense.

With only 50,000 examples on the 3-digits case (3x less) I get 0.97 accuracy after 60 epochs (still underfitting). It's very reasonable.

File: examples/addition_rnn.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+Theoretically it introduces shorter term dependencies between source and target.
+
+Two digits inverted:
++ One layer JZS1 (128 HN) with 55 iterations = 99% train/test accuracy
+Three digits inverted:
++ One layer JZS1 (128 HN) with 19 iterations = 99% train/test accuracy
+Four digits inverted:
++ One layer JZS1 (128 HN) with 20 iterations = 99% train/test accuracy
+Five digits inverted:
++ One layer JZS1 (128 HN) with 28 iterations = 99% train/test accuracy
```
Comment (fchollet) [inline examples/addition_rnn.py]: The TRAINING_SIZE parameter was 800,000 in all these cases?

Comment (fchollet) [inline examples/addition_rnn.py]: Importantly: in each case, how many *unique* examples were generated? (e.g. 150k in the 3 digits case)

**Comment by Smerity** [comment]
+ 2 digits ≅ 5k
+ 3 digits ≅ 150k
+ 4 digits ≅ 390k
+ 5 digits ≅ 550k

To generate a random sample, we first pick a random length from [1, DIGITS] and then produce a number. This is following the "Mixed" learning strategy from [Learning to Execute](http://arxiv.org/pdf/1410.4615v3.pdf) paper.  This is why even at 5 digits there's a fair number of discarded samples. From the paper: "The Mixed strategy uses a balanced mixture of easy and difficult examples, so at every point during training, a sizable fraction of the training samples will have the appropriate difficulty for the RNN."

**Comment by fchollet** [comment]
Merged, with two slight changes:
- make the actual number of training samples match `TRAINING_SIZE`
- add colors to the console output, because I couldn't tell the 'success' character from the 'failure' character
